src/state_machine.zig

```diff
@@ -1572,20 +1572,32 @@ pub fn StateMachineType(
         fn create_transfer_exists(t: *const Transfer, e: *const Transfer) CreateTransferResult {
             assert(t.id == e.id);
             // The flags change the behavior of the remaining comparisons, so compare the flags first.
-            if (@as(u16, @bitCast(t.flags)) != @as(u16, @bitCast(e.flags))) return .exists_with_different_flags;
+            if (@as(u16, @bitCast(t.flags)) != @as(u16, @bitCast(e.flags))) {
+                return .exists_with_different_flags;
+            }
+            // We know that the flags are the same.
+            assert(t.pending_id == 0 and e.pending_id == 0);
+
             if (t.debit_account_id != e.debit_account_id) {
                 return .exists_with_different_debit_account_id;
             }
             if (t.credit_account_id != e.credit_account_id) {
                 return .exists_with_different_credit_account_id;
             }
-            if (t.amount != e.amount) return .exists_with_different_amount;
-            assert(t.pending_id == 0 and e.pending_id == 0); // We know that the flags are the same.
+            // If the accounts are the same, the ledger must be the same.
+            assert(t.ledger == e.ledger);
+
+            // Balancing transfers with `amount=0` must not raise `exists_with_different_amount`.
+            const balancing_transfer = t.amount == 0 and
+                (t.flags.balancing_debit or t.flags.balancing_credit);
+            if (t.amount != e.amount and !balancing_transfer) {
+                return .exists_with_different_amount;
+            }
```

Comment:
Balancing transfers support nonzero amounts as well: a balancing transfer with nonzero amount `A` will move the largest possible amount `X` such that `0 < X ≤ A`. To be consistent, we should consider that case too.

What do you think of this approach?:
If the transfer is created with amount `X` (that is, it actually move `X` -- `A` may have been zero), and then retried with amount `Y`, then the retry should return:

- if `Y=0`: `exists`
- if `Y≥X`: `exists`
- if `Y<X`: `exists_with_different_amount`